clang/lib/Driver/Driver.cpp

```diff
@@ -1066,19 +1083,125 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
   // -ffreestanding cannot be used with -fsycl
   argSYCLIncompatible(options::OPT_ffreestanding);
 
+  llvm::StringMap<llvm::DenseSet<StringRef>> DerivedArchs;
+  llvm::StringMap<StringRef> FoundNormalizedTriples;
   llvm::SmallVector<llvm::Triple, 4> UniqueSYCLTriplesVec;
-
+  // StringSet to contain SYCL target triples.
+  llvm::StringSet<> SYCLTriples;  
+  // If the user specified --offload-arch, deduce the offloading
+  // target triple(s) from the set of architecture(s).
+  // Create a toolchain for each valid triple.
+  // We do not support SYCL offloading if any of the inputs is a
+  // .cu (for CUDA type) or .hip (for HIP type) file.
   if (IsSYCL) {
-    addSYCLDefaultTriple(C, UniqueSYCLTriplesVec);
+    if(C.getInputArgs().hasArg(options::OPT_offload_arch_EQ) && !IsHIP &&
+    !IsCuda) {
+      // SYCL offloading to AOT Targets with '--offload-arch'
+      // is currently enabled only with '--offload-new-driver' option.
+      // Emit a diagnostic if '--offload-arch' is invoked without
+      // '--offload-new driver' option.
+      if (!C.getInputArgs().hasFlag(options::OPT_offload_new_driver,
+                                    options::OPT_no_offload_new_driver, false)) {
+        Diag(clang::diag::err_drv_sycl_offload_arch_new_driver);
```

Comment:
This diagnostic is not necessary for upstream.  The default in upstream in the new offloading model.